### ISSUE TITLE
Able to Pickpocket Equipped Armor, Clothing and Weapons without Requisite Perks

### ISSUE THREAD

**Issue by Velgus** [SniffleMan/QuickLootRE#14]
**Issue:** Currently you can attempt to pickpocket items that NPCs have equipped from the QuickLoot menu without any restrictions.

Normally, you must have the _Misdirection_ perk to attempt to pickpocket equipped weapons, and the _Perfect Touch_ perk to attempt to pickpocked equipped armor/clothing.

**Reproduction:**
1. Ensure you _do not_ have the _Misdirection_ or _Perfect Touch_ perks.
2. Find a (non-hostile) NPC and target them.
3. Switch to sneak mode and view what you are allowed to pickpocket.

**Image:** https://i.imgur.com/e7yCa6O.png

**Comment by SniffleMan** [comment]
Should be fixed in 1.7.4

**Comment by Velgus** [comment]
So the fix still appears to have introduced the following issues:

- While visually hiding the equipped items, they can still be looted by pressing "E" on the empty pickpocket QuickLoot menu.
- The equipped items are also not appearing when the NPC is dead (and the above issue applies as well). All items should be visible when the NPC is dead.
- Pressing E for on said _empty_ pickpocket **or** loot (when the NPC is dead) QuickLoot menu appears to cause crashes after a couple presses (crash may be being triggered by looting equipped "armors"/"clothing", as I noticed in my test case the weapon disappeared from living person/corpse first).
- Pressing "R" from the full inventory of a dead NPC or "F" from the QuickLoot menu of a dead NPC seemed to now cause a crash - possibly related to the above crash. That said, this crash did not occur when I re-tested on a brand new (clean) save - the one I was testing on before was a bit dirty with some mods having been shuffled/removed (the previous crash occured in both cases though).
-Neither of the above crashes appear to happen on any other container type except NPCs, and they also are not triggered by looting 'normally' (without "Take All") from the full inventory menu.

The above issues aside, the detection of the Misdirection and Perfect Touch perks appears to be working. None of the above issues arise if you have 'both' of those perks.

As a side note, I noticed a 'small' UI fix you could make while testing this. When pickpocketing, you actually don't have the option to "Take All" from the regular loot menu. This is true for the QuickLoot menu as well, but perhaps (like the regular loot menu) you could make the "Take All" button prompt disappear for pickpocketing.

**Comment by SniffleMan** [comment]
Should all be fixed in 1.7.5

**Comment by Velgus** [comment]
So it's mostly fixed, but there seems to be an issue of de-sync between what you can 'see' when pickpocketing, and what you can loot. It no longer allows you to pickpocket a 'visually' empty person (so that's fixed), but for instance:

1. Person has lockpicks, salt piles, and clothing. I cannot see clothing because I do not have Perfect Touch.
2. Pickpocket lockpicks - works.
3. Attempt to pickpocket salt piles - it pickpockets the persons hat, despite having the salt piles selected (and the hat not being visible in the QuickLoot menu.

Also, unfortunately I seem to be getting a bit of an unpredictable crashing issue that I haven't been able to isolate with this update. It's happened twice during testing, but it seems to be with different NPCs with wildly varying circumstances. The only consistency between the crashes, is they happened 'immediately' after I entered sneak while having the NPC targeted.

**Comment by Velgus** [comment]
Upon brief testing (with the same NPC who had the specific scenario above) it appears to be fixed. I noticed one other small bug related to lockpicking, but I'll open a new issue for it, as it's not really related. Thank you very much!

**Comment by Velgus** [comment]
Unfortunately after further testing, I spoke too soon.

There still appears to be some sort of indexing error, where I will loot a different item than I have selected (though it no longer causes crashes at all, so that's good).

The same kind of issue as mentioned before still happens, where I will be looting an NPC's items, and at some point, it will switch to looting armors/clothing/weapons, even without the appropriate perks. Visually the perks are fixed, but it still seems to select the items in the background.

Additionally, pickpocketing with QuickLoot is no longer detected as a crime currently (or rather, maybe it is, but ALL pickpocketing chances are locked at 100%, and nothing gets detected even if standing right in front of the NPC being pickpocketed).

**Comment by SniffleMan** [comment]
Are you sure this is still present in 1.7.8? I can't get it to happen

**Comment by Velgus** [comment]
It seems dependent on what's in the NPC's inventory, and possibly sorting rules. I can reproduce it with Jon Battleborn 100% of the time in my game with the default QuickLoot sorting rules. It works up until the Letter from Olfina, at which point it proceeds to loot his weapon and armor.

Before: https://i.imgur.com/NQ9TZJE.jpg
After: https://i.imgur.com/DxG9Dng.jpg

**Comment by SniffleMan** [comment]
See if 1.7.10 fixes it

**Comment by Velgus** [comment]
That seems to have fixed it! I can no longer reproduce with Jon Battleborn, and I ran around town cleaning a bunch of other NPCs contents with no issue. I also re-added the perks, and their functionality is working as normal as well. Thank you!

**Comment by Velgus** [comment]
So this bug is active again as of the latest version (1.15.4). I'm able to loot all equipped items without the perks.

Screenshot: https://i.imgur.com/ueB4ica.jpg

These are my settings: 
[QuickLootRE.zip](https://github.com/SniffleMan/QuickLootRE/files/2676295/QuickLootRE.zip)

**Comment by Velgus** [comment]
Whatever you did in 1.15.5 seems to have fixed it for 'armors', but the problem still exists for weapons.

**Comment by SniffleMan** [comment]
Should be fixed in 1.15.6

**Comment by Velgus** [comment]
It now crashes when trying to open QuickLoot to pickpocket targets with those item types equipped (Daggers and Greatswords).

[Crash_2018_12_13_22-12-1.txt](https://github.com/SniffleMan/QuickLootRE/files/2678846/Crash_2018_12_13_22-12-1.txt)

**Comment by SniffleMan** [comment]
Can you replicate on a vanilla NPC?

**Comment by Velgus** [comment]
Actually I can't, seems to be a problem with that NPC. I tried loading a previous save (before I had ever entered a cell with that NPC) and it still crashed. What's weird is she worked fine (minus the ability to pickpocket her equipped sword) prior to 1.15.6.

**Comment by Velgus** [comment]
I'm curious if it could be caused by the use of the Items\Item\COED - Extra Data field. Certain mods use that feature (I know Inigo uses it, and I think Lucien use it as well).

![image](https://user-images.githubusercontent.com/4391565/49987917-ec6b5800-ff29-11e8-9713-a691ba54b5ae.png)

**Comment by SniffleMan** [comment]
I downloaded interesting NPC's and I can't reproduce the crash

**Comment by Velgus** [comment]
Hmm strange. I have no other mods affecting that NPC in specific (Viranya). You tried it with Viranya in specific I take it (in the Hall of the Vigilant, "coc hallofthevigilant01" or "player.moveto XX005a03")?

Like I mentioned, when I removed her COED data in a patch, it didn't cause a crash. Do you know what that data is used for? Perhaps I have another mod affecting that data.

**Comment by SniffleMan** [comment]
See if 1.15.6.1 helps

**Comment by Velgus** [comment]
Yup, 1.15.6.1 fixed it (without having removed the COED data).

**Comment by Velgus** [comment]
Tested and fixed with 1.15.7 release.
